---
title: "Gene Name Error Monthly Report"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 5
    fig_height: 5
theme: cosmo
---

Source: https://github.com/markziemann/GeneNameErrors2020

View the reports: http://ziemann-lab.net/public/gene_name_errors/

## Intro

Gene name errors result when data are imported improperly into MS Excel and other
spreadsheet programs (Zeeberg et al, 2004).
Certain gene names like MARCH3, SEPT2 and DEC1 are converted into date format.
These errors are surprisingly common in supplementary data files in the field of genomics (Ziemann et al, 2016).
This could be considered a small error because it only affects a small number of genes,
however it is symptomtic of poor data processing methods.
The purpose of this script is to identify gene name errors present in supplementary files of PubMed 
Central articles in the previous month.

```{r,libs}

library("XML")
library("jsonlite")
library("xml2")
library("reutils")
library("readxl")
library("RCurl")

```

## Get PMC IDs

Here I will be getting PubMed Central IDs for the previous month.

Start with figuring out the date to search PubMed Central.

```{r,month}

CURRENT_MONTH=format(Sys.time(), "%m")
CURRENT_YEAR=format(Sys.time(), "%Y")

if (CURRENT_MONTH == "01") {
  PREV_YEAR=as.character(as.numeric(format(Sys.time(), "%Y"))-1)
  PREV_MONTH="12"
} else {
  PREV_YEAR=CURRENT_YEAR
  PREV_MONTH=as.character(as.numeric(format(Sys.time(), "%m"))-1)
}

DATE=paste(PREV_YEAR,"/",PREV_MONTH,sep="")
DATE

```

Let's see how many PMC IDs we have in the past month.

```{r,pmc}

QUERY ='((genom*[Abstract]))'

ESEARCH_RES <- esearch(term=QUERY, db = "pmc", rettype = "uilist", retmode = "xml", retstart = 0, 
  retmax = 5000000, usehistory = TRUE, webenv = NULL, querykey = NULL, sort = NULL, field = NULL, 
  datetype = NULL, reldate = NULL, mindate = DATE, maxdate = DATE)

pmc <- efetch(ESEARCH_RES,retmode="text",rettype="uilist",outfile="pmcids.txt")
pmc <- read.table(pmc)
pmc <- paste("PMC",pmc$V1,sep="")
NUM_ARTICLES=length(pmc)
NUM_ARTICLES
writeLines(pmc,con="pmc.txt")

```

## Run the screen

Now run the bash script. 
Note that false positives can occur (~1.5%) and these results have not been verified by a human.

Here are some definitions:

* NUM_XLS = Number of supplementary Excel files in this set of PMC articles.

* NUM_XLS_ARTICLES = Number of articles matching the PubMed Central search which have supplementary Excel files.

* GENELISTS = The gene lists found in the Excel files.
Each Excel file is counted once even it has multiple gene lists.

* NUM_GENELISTS = The number of Excel files with gene lists.

* NUM_GENELIST_ARTICLES = The number of PMC articles with supplementary Excel gene lists. 

* ERROR_GENELISTS = Files suspected to contain gene name errors.
The dates and five-digit numbers indicate transmogrified gene names.

* NUM_ERROR_GENELISTS = Number of Excel gene lists with errors.

* NUM_ERROR_GENELIST_ARTICLES = Number of articles with supplementary Excel gene name errors.

* ERROR_PROPORTION = This is the proportion of articles with Excel gene lists that have errors.

```{r,screen}

system("./gene_names.sh pmc.txt")
results <- readLines("results.txt")

XLS <- results[grep("XLS",results,ignore.case=TRUE)]
NUM_XLS = length(XLS)
NUM_XLS

NUM_XLS_ARTICLES = length(unique(sapply(strsplit(XLS," "),"[[",1)))
NUM_XLS_ARTICLES

GENELISTS <- XLS[lapply(strsplit(XLS," "),length)>2]
#GENELISTS

NUM_GENELISTS <- length(unique(sapply(strsplit(GENELISTS," "),"[[",2)))
NUM_GENELISTS

NUM_GENELIST_ARTICLES <- length(unique(sapply(strsplit(GENELISTS," "),"[[",1)))
NUM_GENELIST_ARTICLES

ERROR_GENELISTS <- XLS[lapply(strsplit(XLS," "),length)>3]
#ERROR_GENELISTS

NUM_ERROR_GENELISTS = length(ERROR_GENELISTS)
NUM_ERROR_GENELISTS

GENELIST_ERROR_ARTICLES <- unique(sapply(strsplit(ERROR_GENELISTS," "),"[[",1))
GENELIST_ERROR_ARTICLES

NUM_ERROR_GENELIST_ARTICLES <- length(GENELIST_ERROR_ARTICLES) 
NUM_ERROR_GENELIST_ARTICLES

ERROR_PROPORTION = NUM_ERROR_GENELIST_ARTICLES / NUM_GENELIST_ARTICLES
ERROR_PROPORTION

```

## Look at the errors detected

Here you can have a look at all the gene lists detected in the past month, as well as those with errors.
The dates are obvious errors, these are commonly dates in September, March, December and October.
The five-digit numbers represent dates as they are encoded in the Excel internal format.
The five digit number is the number of days since 1900.
If you were to take these numbers and put them into Excel and format the cells as dates, then these will also mostly map to dates in September, March, December and October.

```{r,errors2}

#GENELISTS

ERROR_GENELISTS

```

## In depth look at the errors

Let's investigate the errors in more detail.

```{r,errors,fig.height=8,fig.width=8}

# By species
SPECIES <- sapply(strsplit(ERROR_GENELISTS," "),"[[",3)
table(SPECIES)
par(mar=c(5,12,4,2))
barplot(table(SPECIES),horiz=TRUE,las=1)
par(mar=c(5,5,4,2))

# Number of affected Excel files per paper
DIST <- table(sapply(strsplit(ERROR_GENELISTS," "),"[[",1))
DIST
summary(as.numeric(DIST))
hist(DIST,main="Number of affected Excel files per paper")

# PMC Articles with the most errors
DIST_DF <- as.data.frame(DIST)
DIST_DF <- DIST_DF[order(-DIST_DF$Freq),,drop=FALSE]
head(DIST_DF,20)
MOST_ERR_FILES = as.character(DIST_DF[1,1])
MOST_ERR_FILES

# Number of errors per paper
NERR <- as.numeric(sapply(strsplit(ERROR_GENELISTS," "),"[[",4))
names(NERR) <- sapply(strsplit(ERROR_GENELISTS," "),"[[",1)
NERR <-tapply(NERR, names(NERR), sum)
NERR
hist(NERR,main="number of errors per PMC article")
NERR_DF <- as.data.frame(NERR)
NERR_DF <- NERR_DF[order(-NERR_DF$NERR),,drop=FALSE]
head(NERR_DF,20)
MOST_ERR = rownames(NERR_DF)[1]
MOST_ERR

```

## Journals affected

```{r,journals,fig.width=8,fig.height=8}

GENELIST_ERROR_ARTICLES <- gsub("PMC","",GENELIST_ERROR_ARTICLES)

### JSON PARSING is more reliable than XML
ARTICLES <- esummary( GENELIST_ERROR_ARTICLES , db="pmc" , retmode = "json"  )
ARTICLE_DATA <- reutils::content(ARTICLES,as= "parsed")
ARTICLE_DATA <- ARTICLE_DATA$result
ARTICLE_DATA <- ARTICLE_DATA[2:length(ARTICLE_DATA)]
JOURNALS <- unlist(lapply(ARTICLE_DATA,function(x) {x$fulljournalname} ))
JOURNALS_TABLE <- table(JOURNALS)
JOURNALS_TABLE <- JOURNALS_TABLE[order(-JOURNALS_TABLE)]

length(JOURNALS_TABLE)

NUM_JOURNALS=length(JOURNALS_TABLE)

par(mar=c(5,25,4,2))
barplot(head(JOURNALS_TABLE,10), horiz=TRUE, las=1, 
  xlab="Articles with gene name errors in supp files",
  main="Top journals this month")

```

## Journal of the month winner

Congrats to our Journal of the Month winner!

```{r,journal_winner}

JOURNAL_WINNER <- names(head(JOURNALS_TABLE,1))
JOURNAL_WINNER

```

## Paper of the month winners

There are two categories:

* Paper with the most suplementary files affected by gene name errors (MOST_ERR_FILES)

* Paper with the most gene names converted to dates (MOST_ERR)

Sometimes, one paper can win both categories.
Congrats to our winners.

### Paper with most files affected

```{r,most_err_files}

MOST_ERR_FILES <- gsub("PMC","",MOST_ERR_FILES)
ARTICLES <- esummary( MOST_ERR_FILES , db="pmc" , retmode = "json"  )
ARTICLE_DATA <- reutils::content(ARTICLES,as= "parsed")
ARTICLE_DATA <- ARTICLE_DATA[2]
ARTICLE_DATA

```

### Paper with most date conversions

```{r,most_err}

MOST_ERR <- gsub("PMC","",MOST_ERR)
ARTICLE_DATA <- esummary(MOST_ERR,db = "pmc" , retmode = "json" )
ARTICLE_DATA <- reutils::content(ARTICLE_DATA,as= "parsed")
ARTICLE_DATA

```



## Trend info

To plot the trend over the past 6-12 months.

```{r,trend1,fig.height=8,fig.width=8}

url <- "https://ziemann-lab.net/public/gene_name_errors/"
doc <- htmlParse(url)
links <- xpathSApply(doc, "//a/@href")
links <- links[grep("html",links)]

listing <- htmlParse( getURL(url, ftp.use.epsv = FALSE, dirlistonly = TRUE) )
listing <- xpathSApply(listing, "//a/@href")
listing <- listing[grep("html",listing)]

unlink("online_files/",recursive=TRUE)

dir.create("online_files")

sapply(listing, function(mylink) { 
  download.file(paste(url,mylink,sep=""),destfile=paste("online_files/",mylink,sep=""))  
} )

myfilelist <- list.files("online_files/",full.names=TRUE)

trends <- sapply(myfilelist,  function(myfilename) {

  x <- readLines(myfilename)

  # Num XL gene list articles
  NUM_GENELIST_ARTICLES <- x[grep("NUM_GENELIST_ARTICLES",x)[3]+1]
  NUM_GENELIST_ARTICLES <- sapply(strsplit(NUM_GENELIST_ARTICLES," "),"[[",3)
  NUM_GENELIST_ARTICLES <- sapply(strsplit(NUM_GENELIST_ARTICLES,"<"),"[[",1)
  NUM_GENELIST_ARTICLES <- as.numeric(NUM_GENELIST_ARTICLES)

  # number of affected articles
  NUM_ERROR_GENELIST_ARTICLES <- x[grep("NUM_ERROR_GENELIST_ARTICLES",x)[3]+1]
  NUM_ERROR_GENELIST_ARTICLES <- sapply(strsplit(NUM_ERROR_GENELIST_ARTICLES," "),"[[",3)
  NUM_ERROR_GENELIST_ARTICLES <- sapply(strsplit(NUM_ERROR_GENELIST_ARTICLES,"<"),"[[",1)
  NUM_ERROR_GENELIST_ARTICLES <- as.numeric(NUM_ERROR_GENELIST_ARTICLES)

  # Error proportion
  ERROR_PROPORTION <- x[grep("ERROR_PROPORTION",x)[3]+1]
  ERROR_PROPORTION <- sapply(strsplit(ERROR_PROPORTION," "),"[[",3)
  ERROR_PROPORTION <- sapply(strsplit(ERROR_PROPORTION,"<"),"[[",1)
  ERROR_PROPORTION <- as.numeric(ERROR_PROPORTION)

  # number of journals
  NUM_JOURNALS <- x[grep('JOURNALS_TABLE',x)[3]+1]
  NUM_JOURNALS <- sapply(strsplit(NUM_JOURNALS," "),"[[",3)
  NUM_JOURNALS <- sapply(strsplit(NUM_JOURNALS,"<"),"[[",1)
  NUM_JOURNALS <- as.numeric(NUM_JOURNALS)
  NUM_JOURNALS

  res <- c(NUM_GENELIST_ARTICLES,NUM_ERROR_GENELIST_ARTICLES,ERROR_PROPORTION,NUM_JOURNALS)

  return(res)
})

colnames(trends) <- sapply(strsplit(colnames(trends),"_"),"[[",3)
colnames(trends) <- gsub(".html","",colnames(trends))
trends <- as.data.frame(trends)
rownames(trends) <- c("NUM_GENELIST_ARTICLES","NUM_ERROR_GENELIST_ARTICLES","ERROR_PROPORTION","NUM_JOURNALS")
trends <- t(trends)
trends <- as.data.frame(trends)

CURRENT_RES <- c(NUM_GENELIST_ARTICLES,NUM_ERROR_GENELIST_ARTICLES,ERROR_PROPORTION,NUM_JOURNALS)

trends <- rbind(trends,CURRENT_RES)
paste(CURRENT_YEAR,CURRENT_MONTH,sep="-")
rownames(trends)[nrow(trends)] <- paste(CURRENT_YEAR,CURRENT_MONTH,sep="-")

plot(trends$NUM_GENELIST_ARTICLES, xaxt = "n" , type="b" , main="Number of articles with Excel gene lists per month",
 ylab="number of articles", xlab="month")
axis(1, at=1:nrow(trends), labels=rownames(trends))

plot(trends$NUM_ERROR_GENELIST_ARTICLES, xaxt = "n" , type="b" , main="Number of articles with gene name errors per month",
 ylab="number of articles", xlab="month")
axis(1, at=1:nrow(trends), labels=rownames(trends))

plot(trends$ERROR_PROPORTION, xaxt = "n" , type="b" , main="Proportion of articles with Excel gene list affected by errors",
 ylab="proportion", xlab="month")
axis(1, at=1:nrow(trends), labels=rownames(trends))

plot(trends$NUM_JOURNALS, xaxt = "n" , type="b" , main="Number of journals with affected articles",
 ylab="number of journals", xlab="month")
axis(1, at=1:nrow(trends), labels=rownames(trends))

unlink("online_files/",recursive=TRUE)

```

## References

1. Zeeberg, B.R., Riss, J., Kane, D.W. et al. Mistaken Identifiers: Gene name errors can be introduced inadvertently when using Excel in bioinformatics. BMC Bioinformatics 5, 80 (2004). https://doi.org/10.1186/1471-2105-5-80

2. Ziemann, M., Eren, Y. & El-Osta, A. Gene name errors are widespread in the scientific literature. Genome Biol 17, 177 (2016). https://doi.org/10.1186/s13059-016-1044-7

## SessionInfo

```{r,sessioninfo}

sessionInfo()

```
